data_model/src/query/predicate/predicate_atoms/block.rs

```diff
@@ -19,53 +21,254 @@ use crate::{
-        match *self {}
+    fn applies(&self, input: &BlockHeader) -> bool {
+        match self {
+            BlockHeaderPredicateBox::Hash(hash) => hash.applies(&input.hash()),
+        }
     }
 }
 
 /// A predicate that can be applied to a [`SignedBlock`].
 #[derive(Debug, Clone, PartialEq, Eq, Decode, Encode, Deserialize, Serialize, IntoSchema)]
 pub enum SignedBlockPredicateBox {
-    // nothing here yet
+    // projections
+    /// Checks if a predicate applies to the header of the block.
+    Header(BlockHeaderPredicateBox),
 }
 
 impl_predicate_box!(SignedBlock: SignedBlockPredicateBox);
 
 impl EvaluatePredicate<SignedBlock> for SignedBlockPredicateBox {
-    fn applies(&self, _input: &SignedBlock) -> bool {
-        match *self {}
+    fn applies(&self, input: &SignedBlock) -> bool {
+        match self {
+            SignedBlockPredicateBox::Header(header) => header.applies(input.header()),
+        }
+    }
+}
+
+/// A predicate that can be applied to a [`HashOf<SignedTransaction>`].
+#[derive(Debug, Clone, PartialEq, Eq, Decode, Encode, Deserialize, Serialize, IntoSchema)]
+pub enum TransactionHashPredicateBox {
+    // object-specific predicates
+    /// Checks if the input is equal to the expected value.
+    Equals(HashOf<SignedTransaction>),
+}
+
+impl_predicate_box!(HashOf<SignedTransaction>: TransactionHashPredicateBox);
+
+impl EvaluatePredicate<HashOf<SignedTransaction>> for TransactionHashPredicateBox {
+    fn applies(&self, input: &HashOf<SignedTransaction>) -> bool {
+        match self {
+            TransactionHashPredicateBox::Equals(hash) => input == hash,
+        }
+    }
+}
+
+/// A predicate that can be applied to a [`SignedTransaction`]
+#[derive(Debug, Clone, PartialEq, Eq, Decode, Encode, Deserialize, Serialize, IntoSchema)]
+pub enum SignedTransactionPredicateBox {
+    // projections
+    /// Checks if a predicate applies to the hash of the signed transaction.
+    Hash(TransactionHashPredicateBox),
+    /// Checks if a predicate applies to the authority of the signed transaction.
+    Authority(AccountIdPredicateBox),
+}
+
+impl_predicate_box!(SignedTransaction: SignedTransactionPredicateBox);
+
+impl EvaluatePredicate<SignedTransaction> for SignedTransactionPredicateBox {
+    fn applies(&self, input: &SignedTransaction) -> bool {
+        match self {
+            SignedTransactionPredicateBox::Hash(hash) => hash.applies(&input.hash()),
+            SignedTransactionPredicateBox::Authority(authority) => {
+                authority.applies(input.authority())
+            }
+        }
+    }
+}
+
+// TODO: maybe we would want to have a generic `Option` predicate box & predicate
+/// A predicate that can be applied to an [`Option<TransactionRejectionReason>`]
+#[derive(Debug, Clone, PartialEq, Eq, Decode, Encode, Deserialize, Serialize, IntoSchema)]
+pub enum TransactionErrorPredicateBox {
+    // object-specific predicates
+    /// Checks if there was an error while applying the transaction.
+    IsSome,
```

Comment:
Should we add `IsNone`? to be able to query successful transactions or we can negate `IsSome`?

Reply:
You already can do this by negating `IsSome`. The prototype provides an `is_none` helper function that does the negation for you https://github.com/hyperledger/iroha/blob/5b63359151f6185a29ce3b59c4a6e96d6355b955/data_model/src/query/predicate/prototypes/block.rs#L119-L123